lib/puppet/provider/nameservice.rb

```diff
@@ -288,5 +288,25 @@ def set(param, value)
       raise Puppet::Error, "Could not set #{param} on #{@resource.class.name}[#{@resource.name}]: #{detail}", detail.backtrace
     end
   end
+
+  # From overriding Puppet::Property#insync?
+  # Ruby Etc::getpwnam < 2.1.0 always returns a struct with binary encoded
+  # string values, and >= 2.1.0 will return binary encoded strings for values
+  # incompatible with current locale charset. Convert "should" value to binary
+  # if "current" value is, to avoid unnecessary property syncs and concatenation
+  # of strings with incompatible encodings. (PUP-6777)
+  #
+  # * "should" parameter will always be a string with one possible (strictly
+  # speaking) exception, which is the boolean value true. In current semantics,
+  # this would convert to a string. For strict backwards-compatibility, convert
+  # it to a string if this is the case.
+  #
+  # return basic string comparison after encoding (same as
+  # Puppet::Property#property_matches)
+  def comments_insync?(current, should)
+    should = should.to_s if should == true
```

Comment:
yeah - the to_s is almost always going to be unnecessary (unless we go with `@should` as you suggest below). The only time this will be the case is if the user has provided the literal value `true` as the value to the `comment` property, ie
```ruby
user { 'foo' : comment => true }
```
before this change, that `true` would be cast to a string, so I wanted to ensure that stays the same way.